File: CpuA32/disassembler_tool.cc
// (c) Robert Muth - see LICENSE for more info

#include "CpuA32/disassembler.h"
#include "Util/assert.h"

#include <cstdlib>
#include <iostream>
#include <string_view>

using namespace cwerg::a32;

int main(int argc, char* argv[]) {
  if (argc <= 1) {
    std::cout << "no command specified\n";
    return 1;
  }
  if (std::string_view("disass") == argv[1]) {
    for (unsigned i = 2; i < argc; ++i) {
      const uint32_t data = strtoul(argv[i], 0, 16);
      Ins ins;
      if (!DecodeIns(&ins, data)) {
        std::cout << "could not find opcode for: " << std::hex << data << "\n";
        continue;
      }
      char buffer[128];
      std::cout << argv[i] << " " << ins.opcode->enum_name;
      std::string_view sep = " ";
      for (unsigned i = 0; i < ins.opcode->num_fields; ++i) {
        RenderOperandStd(buffer, *ins.opcode, ins.operands[i],
                         ins.opcode->fields[i]);
        std::cout << sep << buffer;
        sep = ", ";
      }
      std::cout << "\n";
      // check that the assembler works - this is not strictly
      // necessary but useful for debuggging the assembler
      const uint32_t data2 = EncodeIns(ins);
      if (data != data2) {
        std::cout << "Disassembler failure " << std::hex << data << " vs "
                  << data2 << "\n";
        return 1;
      }
      ASSERT(data == data2, "");
    }
  } else if (std::string_view("benchmark") == argv[1]) {
    // The first 1<<28 (256M) bit patterns exercise all opcodes with predicate
    // "eq"
    uint32_t num_bad = 0;
    for (uint32_t data = 0; data < (1 << 28); ++data) {
      Ins ins;
      if (DecodeIns(&ins, data)) {
        const uint32_t data2 = EncodeIns(ins);
        if (data != data2) {
          if (ins.opcode->fields[2] != OK::IMM_0_7_8_11 &&
              ins.opcode->fields[3] != OK::IMM_0_7_8_11) {
            std::cout << "Disassembler failure " << std::hex << data << " vs "
                      << data2 << ": ";
            char buffer[128];
            std::cout << ins.opcode->enum_name;
            std::string_view sep = " ";
            for (unsigned i = 0; i < ins.opcode->num_fields; ++i) {
              RenderOperandStd(buffer, *ins.opcode, ins.operands[i],
                               ins.opcode->fields[i]);
              std::cout << sep << buffer;
              sep = ", ";
            }
            std::cout << "\n";

            return 1;
          }
          continue;
        }
      } else {
        ++num_bad;
      }
      if (data % 10000000 == 0) {
        std::cout << num_bad << "/" << data << "\n";
      }
    }
    std::cout << "unsupported opcodes: " << std::dec << num_bad << "\n";
    return 0;
  } else {
    std::cout << "unknown command specified: " << argv[1] << "\n";
    return 1;
  }
}

File: CpuA32/disassembler_tool.py
#!/usr/bin/python3

import sys

from CpuA32 import symbolic
from CpuA32 import opcode_tab as a32


def disass(data):
    ins = a32.Disassemble(data)
    if ins.opcode is None:
        print(f"could not disassemble {data:x}")
        return
    enum_name, operands_str = symbolic.InsSymbolize(ins)

    print(f"{data:08x} {enum_name} {', '.join(operands_str)}")
    for f, o, o_str in zip(ins.opcode.fields, ins.operands, operands_str):
        print(f"    {f.name:25} {o_str:15} ({o})")
    print()
    data2 = a32.Assemble(ins)
    assert data == data2
    ins2 = symbolic.InsFromSymbolized(enum_name, operands_str)
    assert tuple(ins.operands) == tuple(
        ins2.operands), f"{ins.operands} vs {ins2.operands}"


def batch():
    for line in sys.stdin:
        if not line or line.startswith("#"): continue
        data = int(line.split()[0], 16)
        ins = a32.Disassemble(data)
        if ins.opcode is None:
            print(f"could not disassemble {data:x}")
            continue
        enum_name, ops_str = symbolic.InsSymbolize(ins)
        print(f"{data:08x} {enum_name}{' ' if ops_str else ''}{', '.join(ops_str)}")
        data2 = a32.Assemble(ins)
        assert data == data2
        ins2 = symbolic.InsFromSymbolized(enum_name, ops_str)
        assert tuple(ins.operands) == tuple(
            ins2.operands), f"{ins.operands} vs {ins2.operands}"


if __name__ == "__main__":
    if len(sys.argv) > 1:
        if sys.argv[1] == "batch":
            batch()
        else:
            for arg_hex_number in sys.argv[1:]:
                disass(int(arg_hex_number, 16))
